Integrate statuspage.io in the UI

ciyer opened the pull request:
The is PR implements the design specified in #938.
The configuration for this feature is documented in https://github.com/SwissDataScienceCenter/renku/pull/1315.

The dev-values file template is updated in this PR: https://github.com/SwissDataScienceCenter/sdsc-dev-docs/pull/65

# Testing
- if the statuspage.id value is not configured, then you should not see the help/status page
- if the statuspage.id value is set to junk, then you should see a banner explaining that the statuspage is not correctly configured in the help/status page
- if you add a scheduled maintenance, you should a notification banner and details in the help/status page
- if a component is not stable, you should a notification banner and details in the help/status page

## Creating / setting a statuspage id

### Create a statuspage

You can create a free account at [statuspage.io](http://statuspage.io). 

### Setting the statuspage id

By default, no statuspage is configured. You can configure one by filling out the configuration.
```
ui:
  statuspage:
    id: [statuspage id]
```

The `run-telepresence.sh` script sets the statuspage to the dev statuspage id by default. You can override this by setting a STATUSPAGE_ID environment variable.

Finally, you can modify `src/index.js:81` and set the id explicitly. Just change this line to contain your statuspage id:
```const statuspageId = params["STATUSPAGE_ID"];```

The last option is nice for testing since you can change the statuspage id without having to redeploy.

# Screenshots

## Scheduled Maintenance

![image](https://user-images.githubusercontent.com/1196411/88563199-f9bb5480-d031-11ea-9bdf-0b25e66a5f71.png)

## Problems with Service (details)

![image](https://user-images.githubusercontent.com/1196411/88563394-3ab36900-d032-11ea-9506-9372b5222a4a.png)

## Problems with Service (landing page)
![image](https://user-images.githubusercontent.com/1196411/88659485-69cee680-d0d5-11ea-8f09-e5b0627966f3.png)

## Incorrectly configured statuspage

![image](https://user-images.githubusercontent.com/1196411/88792930-4fac0b80-d19c-11ea-9959-c6eefafe9b87.png)

Fix #938

rokroskar commented:
This looks really nice! Does the user get a popup if there is a maintenance announcement that they haven't seen yet? Or do they have to go to "Status" to see that there is a planned or ongoing maintenance/outage? One option would be to have a small "indicator light" somewhere that would quickly give a clue that something is going on.

ciyer commented:
> This looks really nice! Does the user get a popup if there is a maintenance announcement that they haven't seen yet? Or do they have to go to "Status" to see that there is a planned or ongoing maintenance/outage? One option would be to have a small "indicator light" somewhere that would quickly give a clue that something is going on.

This is in progress. Just added a screenshot for the outage announcement that is shown on the landing page. Adding the same for scheduled maintenance.

Once those are done, the only outstanding TODO is to poll periodically for changes.

rokroskar commented:
Would it be useful to have a sample statuspage set up? I _think_ we can have one that is not published but should still allow you to use the API.

ciyer commented:
> Would it be useful to have a sample statuspage set up? I _think_ we can have one that is not published but should still allow you to use the API.

I just set one up myself for testing. The id for the statuspage is a configuration parameter.

rokroskar commented:
Ok perfect. I have a temp one that we set up for limited, but if you have it already then all good.

rokroskar commented on helm-chart/renku-ui/values.yaml:
```diff
@@ -73,3 +73,7 @@ sentry:
   url: ''
   namespace: ''
+
+# If you maintain a statuspage.io page for the status of your RenkuLab instance, put the id for it here.
+statuspage:
+  id: ''
```
```suggestion
##  If you maintain a statuspage.io page for the status of your RenkuLab instance, put the id for it here.
# statuspage:
#  id: 
```

ciyer replied:
Done.

rokroskar replied:
Actually I think my suggestion will break the chart rendering with default values 🙈  

It should be

```
statuspage:
  id:
```

i.e. the elements should be there but not set to anything.

ciyer replied:
I tested it, and I didn't have any problems with chart rendering in either case. Both "not specified at all" and "specified to empty" work. So it is your call how you want the chart to look.

rokroskar replied:
I think it's better to just leave it null since that's what we have elsewhere and is usually what is done in helm charts.

ciyer replied:
It's the way you requested now.

rokroskar commented on helm-chart/renku-ui/templates/deployment.yaml:
```diff
@@ -34,6 +34,8 @@ spec:
               value: {{ .Values.gatewayUrl | default (printf "%s://%s/api" (include "ui.protocol" .) .Values.global.renku.domain) | quote }}
             - name: WELCOME_PAGE
               value: {{ .Values.welcomePage.text | b64enc | quote }}
+            - name: STATUSPAGE_ID
```
would be better to put this into a conditional and not set it at all if the statuspage id is missing

ciyer replied:
Done.

rokroskar commented on run-telepresence.sh:
```diff
@@ -38,6 +38,8 @@ else
   echo "Warning! your OS has not been tested yet"
 fi
 
+if [ -z "$STATUSPAGE_ID" ]; then STATUSPAGE_ID="5bcn9bqff4qt"; else echo "STATUSPAGE_ID is set to '$STATUSPAGE_ID'"; fi
```
don't put the renkulab id as a default

ciyer replied:
Switched to renku-dev.

ciyer commented on run-telepresence.sh:
```diff
@@ -93,7 +95,9 @@ tee > ./public/config.json << EOF
   "PRIVACY_ENABLED": "false",
-  "TEMPLATES": ${TEMPLATES}
+  "TEMPLATES": ${TEMPLATES},
+  "ANONYMOUS_SESSIONS": "true",
```
Good catch!

ciyer commented on src/index.js:
```diff
@@ -40,9 +40,9 @@ Promise.all([configFetch, privacyFetch]).then(valuesRead => {
-    if (maintenace) {
-      ReactDOM.render(<Maintenance info={maintenace} />, document.getElementById("root"));
+    const maintenance = params["MAINTENANCE"];
+    if (maintenance && maintenance !== "false" && maintenance !== "0") {
```
👍

ciyer commented on src/statuspage/StatuspageAPI.js:
```diff
@@ -0,0 +1,56 @@
+    this.controller = null;
+  }
+
+  async summary() {
```
I wasn't sure what to do about this. It is true that our other API code is in that folder, but those all calls to the gateway; the statuspage API goes directly to statuspage.io, so it made sense to me to keep that code together with the rest of the statuspage support.

If you want to argue for moving it, I am open to it.

lorenzo-cavazzi replied:
I generally like the idea of keeping all the APIs in one place. It makes it easier to find them if something needs to be changed. Otherwise, in this case, I can't imagine re-using the API anywhere else, therefore it's probably not much important where the code is

ciyer commented on src/statuspage/Statuspage.present.js:
```diff
@@ -0,0 +1,275 @@
+      <br />
+      <p>
+        For further information, see <a href={summary.statuspage.page.url}>{summary.statuspage.page.url}</a>.<br />
+        <TimeCaption caption="Status retrieved" time={summary.retrieved_at} endPunctuation=";" />
```
Yes, I was also unsure about how best to handle this. I removed the "Status retrieved" part, but then I found it confusing because it wasn't clear if the information being displayed was retrieved, say two days ago, or just now, so I decided to leave it in.

lorenzo-cavazzi replied:
An idea would be to re-compute the human-readable text every `n` seconds if the provided time hasn't changed in the meanwhile.
That's probably an improvement we can add in the `TimeCaption` component. We can open an issue if we think it's useful, but the impact here is rather limited.

ciyer commented:
> The code looks great, I left just a few minor comments.

Good catches there, thanks. I think I have addressed all your comments.

> I have 2 considerations (it's not necessary to address them in the PR):
> 
> * When I load the home page, I get the warning on top almost immediately, but after a few seconds it goes away for a moment and then it re-appears. It seems that data are fetched again when the user's projects are loaded.

Yes, this is very ugly, and I spent some time trying to figure out how to fix it, but I was not able to find a solution. I will make a follow-up issue. 

> * It would be nice to have the same info also on the landing page for anonymous users.

Good point. The UX should probably be a bit different for anonymous users. I will make a follow-up for this as well.

lorenzo-cavazzi reviewed (approved):
All good! :tada: 

Let me know if it makes sense to create an issue for auto-updating the sentence for the `TimeCaption` component.

ciyer commented:
Follow-up issues:
- #1016: fix flashing
- #1017: added status pages for anonymous users
